Sources/Event Logging (Swift)/TrackEventPersistenceService.swift

```diff
@@ -0,0 +1,82 @@
+#if SWIFT_PACKAGE
+import AutomatticTracksModel
+import AutomatticTracksModelObjC
+#endif
+import CoreData
+
+/// Should eventually replace objC class `TracksEventPersistenceService`.  Unfortunately
+/// Swift Packages do not support mixed source classes so we'll instead need to compose.
+///
+@objc
+public class TrackEventPersistenceService: NSObject {
+    
+    private let managedObjectContext: NSManagedObjectContext
+    
+    @objc
+    public init(managedObjectContext: NSManagedObjectContext) {
+        self.managedObjectContext = managedObjectContext
+        
+        super.init()
+    }
+    
+    /// Increments the retry count for the specified events in batches of
+    @objc
+    public func incrementRetryCountForEvents(_ tracksEvents: [TracksEvent], onComplete completion: ((Error?) -> Void)?) {
+        let uuidStrings = tracksEvents.map { event in
+            event.uuid.uuidString
+        }
+        
+        let batchSize = 500
+        
+        for startIndex in stride(from: 0, to: uuidStrings.count, by: batchSize) {
+            let isLastBatch = startIndex + batchSize >= uuidStrings.count
+            
+            managedObjectContext.perform {
```

Comment:
This is dangerous in my opinion.

It's good in that it allow us to batch the requests to CoreData, but by doing it this way we're effectively allowing other `perform` calls to execute in-between our own `perform` calls, potentially changing the previous behaviour.

I did this because batching sounded like a good idea, but right after finishing with the implementation I wondered if having a single perform block would be better.  We could still keep the batching inside the perform block, although I'm not sure that'd be useful for anything other than memory.

I need some thoughts here.

Reply:
Now that this is no longer happening on the main thread, I think it would be fine to keep it all in a single `perform` block. Batching for the sake of memory might be nice, but is probably less critical.

One way to simplify this could be to use `request.fetchBatchSize` to handle the batching automatically. Note, though, that if you're going to use that, you need to interact with the results as an NSArray, not a Swift array. (This is because it actually returns a special subclass of NSArray that fetches things on-demand in batches, but converting it to a Swift array just immediately reads everything, which triggers all the fetching right away and defeats the point of the batching.)